185495405 custom data importers

ttoomey opened the pull request:
https://www.pivotaltracker.com/story/show/185495405

Import custom data elements, custom services, referrals, unit occupancy, client contacts, client phones,
* per discussion, not importing MCI ids. Those are already done
* per discussion, sticking with smoke tests

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
Merge branch 'release-84' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
Merge branch 'release-84' into 185495405-custom-data-importers

ttoomey commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/client_contacts_loader.rb:
```diff
@@ -0,0 +1,50 @@
+          use: use_value(row),
+          system: row_value(row, field: 'SYSTM').downcase,
+          notes: row_value(row, field: 'NOTES'),
+          value: row_value(row, field: 'VALUE'),
```
Skip all 000s, strip out non-numerics. At least 10 digits, else skip

ttoomey commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_definitions.rb:
```diff
@@ -0,0 +1,62 @@
+        # Enrollment: rental_assistance_end_date
+        { owner_type: 'Hmis::Hud::Enrollment', field_type: :date, key: :rental_assistance_end_date, label: 'Rental Assistance End Date', at_occurrence: true },
+        # Enrollment: esg_allowance_grant_referred
+        { owner_type: 'Hmis::Hud::Enrollment', field_type: :integer, key: :esg_allowance_grant_referred, label: 'ESG Allowance Grant Referred' },
```
should be string

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
Merge branch 'release-84' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
Merge branch 'release-84' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
Merge branch 'release-84' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+        )
+        # not totally sure how to treat these dates
+        if record.assigned_status?
+          record.status_updated_at = parse_date(row_value(row, field: 'ASSIGNED_AT'))
```
`ASSIGNED_AT` should go into `created_at`

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+
+    def referral_scope
+      enrollments = Hmis::Hud::Enrollment.where(data_source: data_source)
+      referral_class.where(enrollment_id: enrollments.select(:id))
```
```suggestion
      referral_class.where(enrollment_id: enrollments.pluck(:id))
```
since we don't need the AR records?

ttoomey replied:
I like to use select rather than pluck since AR turns that into a subquery which is usually (but not always) more efficient.

However in this case I just removed the DS condition since the relation between DS and referral is a little awkward and there shouldn't be any other referrals

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+          record.status_updated_at = parse_date(row_value(row, field: 'ASSIGNED_AT'))
+        else
+          record.status_updated_at = parse_date(row_value(row, field: 'STATUS_UPDATED_AT'))
+        end
```
We're not yet creating new Enrollments for records in `accepted_pending_status`, correct?


I know we already went over the fact that `entry_date` and `relationship_to_hoh` need to be set, and `household_id`. In addition, the EnrollmentCoC col is new in 2024, and required, so we need to set that.

It will be the same for all projects, so this can be made more efficient by pulling the coc code into an instance variable or something. But it's: `project.project_cocs.first.coc_code`.
The expectation is that all projects have exactly 1 ProjectCoC record, all with the same coc_code value, since this is a Single CoC installation.

ttoomey replied:
Good to know, thanks. I populated the coc field on the new enrollments

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+        .relationship_to_hohs
+        .invert.stringify_keys
+      # the field name here is different from the spec
+      @posting_status_map.fetch(row_value(row, field: 'RELATIONSHIP_TO_HOH_ID'))
```
Thanks, missed that col name change, I updated the Google Sheet.

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+    end
+
+    # we don't have enrollment id on the referrals csv so we have to infer it
+    # from the referral.project_id and MCI ID on household member.
```
Ok this part will be updated to use the ENROLLMENT_ID col instead, which should match up with a `Hmis::Hud::Enrollment` record's `EnrollmentID` field (and data source needs to match too).

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,229 @@
+        referral_id = row_value(row, field: 'REFERRAL_ID')
+        referral_class.new(
+          identifier: referral_id,
+          enrollment_id: referral_enrollment_id(referral_id),
```
```suggestion
```

Sorry, this is confusingly named, that's my bad. The `enrollment_id` col stores a reference to the enrollment that the referral _came from_, which only gets set for referrals that are made within the HMIS. For LINK-originating referrals, it's always null.

Instead, the "link" to the enrollment is stored by way of `HouseholdID` column on the `hmis_external_referral_postings` table.
So:
1. For accepted referrals, we ensure that we can find the enrollment by ENROLLMENT_ID, and then store the enrollment's household ID on the referral posting record.
2. For accepted pending referrals, we create an enrollment for each household member, and then store the household ID on the referral posting record.

The reason for using household ids, rather than enrollment ids, is because the household membership my change but it still needs to remain "linked" to the posting.

ttoomey replied:
Thanks for clarifying this! This all sounds familiar, I think I misremembered how it was used 🤦 
I added comments to the model classes to document it and removed the enrollment reference from the referrals.

gigxz reviewed (commented):
I'm still making my way through, but here are some initial comments on the referral pieces!

ttoomey committed:
185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers

organize referral loader class

ttoomey commented:
Thanks for the initial review @gigxz - I've updated the referral importer:
* use the enrollment id for accepted referrals
* create wip enrollments for accepted-pending referrals
* removed unused lookups and reordered the methods for readability

gigxz commented:
@ttoomey this will go into `release-86`, can you retarget? release-84 is closed, and 85 is reserved for ACL work only. thanks!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/project_unit_tracker.rb:
```diff
@@ -0,0 +1,69 @@
+      @household_assignments = {}
+      @assignments = {}
+
+      # we don't check if the unit is occupied, assumption is that all unit occupancies
+      # are deleted before import
```
Added to https://docs.google.com/document/d/1TgQ-Y6-49UwQRnpH_7Pz5gnAYDBuzP5ldeS1YL_3z2I/edit

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/project_unit_tracker.rb:
```diff
@@ -0,0 +1,69 @@
+      return if assignments[enrollment_pk] || unit_type_mper_id.blank?
+
+      unit_pk = unit_pk_for_enrollment_pk(enrollment_pk, unit_type_mper_id)
+      return unless unit_pk
```
So this would happen if we ran out of units to assign, right? Can we log a warning here that includes the enrollment pk and unit type mper id?

gigxz replied:
I'm curious to see if this comes up in the staging test migration.

ttoomey replied:
oh yes, good point, I'll add some logging for this case. Thanks!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+
+    def connection
+      ProjectsImportAttempt.connection
```
I'm a little confused about how we're using this table, can you explain?
Looks like it was set up initially to track the MPER imports – is there a way to distinguish between records from MPER import vs this one? (If we are creating a record..)

ttoomey replied:
Oh yeah, I copied the projects importer originally and that got left behind. The specific class used doesn't really matter as long as it's a warehouse model. I'll change it to be Project.connection to be more clear.

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/emergency_shelter_allowance_grant_loader.rb:
```diff
@@ -0,0 +1,73 @@
+            definition_key: :esg_allowance_grant_received_amount,
+          ),
+          new_cde_record(
+            # FIXME - need to map integer value, mapping not yet provided
```
```suggestion
```

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/emergency_shelter_allowance_grant_loader.rb:
```diff
@@ -0,0 +1,73 @@
+        .pluck(:enrollment_id, :id)
+        .to_h
+      rows.flat_map do |row|
+        enrollment_id = row_value(row, field: 'ENROLLMENTID')
```
Should we pass this flag so we get an error if its missing?
```suggestion
        enrollment_id = row_value(row, field: 'ENROLLMENTID', required: true)
```

gigxz replied:
nevermind I realized its the default 😁

ttoomey replied:
So `required: true` is actually the default for BaseLoader.row_value(). I made it the default since it seemed like most fields were required in the spec.

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb:
```diff
@@ -0,0 +1,135 @@
+
+    def row_value(row, field:, required: true)
+      value = row[field]&.strip&.presence
+      raise "field '#{field}' is missing" if required && !value
```
When this error comes up, will we know which file it's coming from? Hopefully we can see it in the stack trace, just want to confirm. Maybe we should log the whole row so we can more easily find the culprit too.

ttoomey replied:
Good idea, I'll add the full row to the error message here.

So the idea is that CsvFile decorates any exception that occurs as the file is being read (using rows.each) with the relevant filename and row number. That should let us locate the problem row. Does that seem sufficient?

gigxz replied:
Definitely! thanks

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/emergency_shelter_allowance_grant_loader.rb:
```diff
@@ -0,0 +1,73 @@
+      ]
+    end
+
+    def build_records
```
Maybe I'm missing something obvious, but where does this method get called? Other loaders have a `perform`

ttoomey replied:
So this should work as follows:
* This class inherits from CustomDataElementBaseLoader which defines perform
* CustomDataElementBaseLoader.perform calls build_records

It felt convenient and DRY to use inheritance here but it is a little indirect and hard to follow; lmk if I should refactor

gigxz replied:
Oh okay gotcha, thanks!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/deferred_project_unit_occupancy_loader.rb:
```diff
@@ -0,0 +1,45 @@
+      rows.map do |row|
+        unit_id, enrollment_id, start_date = row.fetch_values(:unit_id, :enrollment_id, :start_date)
+        record = model_class.new(unit_id: unit_id, enrollment_id: enrollment_id)
+        record.build_occupancy_period(start_date: start_date || today, user_id: system_user_id)
```
Welp, `user_id` on this table should be the primary key for the `Hmis::User` (`users` table). I'll look for other usages of `system_user_id` and check them. (Edit: this was the only instance i found).

```suggestion
        record.build_occupancy_period(start_date: start_date || today, user_id: Hmis::User.system_user.id)
```


Its confusing, but the rule is:
* If it's a "hud" table, the table name is camelcase and so are the column names. it should have a column `UserID`. In this case, `UserID` maps to the `UserID` column on the `Users` table.
* If it's a "non-hud" table, the table name is snake-cased and so are the column names. it may have a column `user_id`. In this case, `user_id` maps to the `id` column on the `users` table.

This has tripped me up sooo many times, I wish it weren't this way, but it is. cc @eanders we don't have this documented anywhere, do we? we should, def a quirk when onboarding to the codebase (And beyond)

eanders replied:
Yeah, not documented that I know of.  Leave it to HUD to slip in the Users table a few years ago.  I don't have a better solution, not really sure where to document it, other than maybe the user.rb files.

gigxz replied:
Okay. yeah I was thinking something on the "HUD tables" vs "non HUD tables" more broadly in the readme somewhere. To do before onboarding the next contributor 🙂 

Since there are some other "gotchas" there around data source id, uniqueness id "HUD IDs", etc

ttoomey replied:
Great catch! I missed that we were using the pk here. Thanks for auditing the other uses of system_user_id. It's too bad we can't have a FK back to user here, that would be a nice safety check.

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb:
```diff
@@ -0,0 +1,108 @@
+          DateUpdated: parse_date(row_value(row, field: 'DATEUPDATED')),
+          custom_service_type_id: custom_service_type.id,
+          PersonalID: personal_id_by_enrollment_id.fetch(enrollment_id),
+          DateProvided: today, # FIXME - this isn't right?
```
This can be the same as `FAStartDate`, that's how we're doing it for new services.

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb:
```diff
@@ -0,0 +1,108 @@
+        .to_h
+
+      rows.map do |row|
+        enrollment_id = row_value(row, field: 'ENROLLMENTID')
```
Again, should we pass `required` so that we error here rather than at validation

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+
+    FUNDING_SOURCE_MAP = {
+      'Allegheny County' => 'Allegheny County ESG',
+      'City of Pittsburgh' => 'City of Pittsburgh ESG',
+      'State of Pennsylvania' => 'State of Pennsylvania ESG',
```
Did the stars get stripped out somewhere? In the initial file there were these values
```
 '**Allegheny County**' => 'Allegheny County ESG',
  '**City of Pittsburgh**' => 'City of Pittsburgh ESG',
  '**State of Pennsylvania**' => 'State of Pennsylvania ESG',
```

ttoomey replied:
Oh oops, I forgot the asterisks were actually in the data. Fixed now

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:
```diff
@@ -0,0 +1,41 @@
+        .where(data_source: data_source)
+        .pluck(:income_benefits_id, :id, :enrollment_id)
+        .to_h { |income_benefits_id, pk, enrollment_id| [income_benefits_id, [pk, enrollment_id]] }
+      rows.map do |row|
+        benefits_id = row_value(row, field: 'INCOMEBENEFITSID')
+        benefit_pk, enrollment_id = benefit_id_lookup.fetch(benefits_id)
+        raise 'BenefitsID/EnrollmentID mismatch' if enrollment_id != row_value(row, field: 'ENROLLMENTID')
+
+        new_cde_record(
+          value: row_value(row, field: 'FEDERALPOVERTYLEVEL', required: false),
+          definition_key: :federal_poverty_level,
+        ).merge(owner_id: benefit_pk)
+      end
```
I think we can add in an early return here. Looking at the old data file, the `FEDERALPOVERTYLEVEL` has no null values but it does have some `Data not collected` values which we don't need to bother with.

```
rows.map do |row|
   value = row_value(row, field: 'FEDERALPOVERTYLEVEL', required: false)
   return if value.nil? || value == 'Data not collected'
```

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/reason_for_exit_loader.rb:
```diff
@@ -0,0 +1,84 @@
+        exit_id = row_value(row, field: 'ExitID')
+        owner_id, enrollment_id = exit_lookup[exit_id]
+        raise 'ExitID/EnrollmentID mismatch' if enrollment_id != row_value(row, field: 'EnrollmentID')
+
```
Looks like there are some rows that are missing all 3 columns (ReasonForExit, ReasonForExitOther, VoluntaryTermination). It doesn't hurt to create empty CDEs in that case, but it would also be fine to skip those rows.

ttoomey replied:
new_cde_record() returns nil if the provided value is nil so we should already be skipping these records

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,269 @@
+      accepted_pending_status = HmisExternalApis::AcHmis::ReferralPosting.statuses.fetch('accepted_pending_status')
+      enrollment_coc_by_project_id = Hmis::Hud::ProjectCoc
+        .where(data_source: data_source)
+        .pluck(:project_id, :project_coc_id)
```
```suggestion
        .pluck(:project_id, :coc_code)
```

gigxz replied:
EnrollmentCoC 3.16.1 is the coc code itself https://files.hudexchange.info/resources/documents/HMIS-CSV-Format-Specifications-2024.pdf

ttoomey replied:
oh oops, thanks!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,269 @@
+            project_id: project_pk_by_id(project_id),
+          )
+          enrollment
+        end
```
This looks good to me! Thank you for making something complicated so readable!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_requests_loader.rb:
```diff
@@ -0,0 +1,68 @@
+          requested_on: parse_date(row_value(row, field: 'REQUESTED_ON')),
+          needed_by: parse_date(row_value(row, field: 'NEEDED_BY')),
+          # use empty string for non-nullable columns here
+          requestor_name: row_value(row, field: 'REQUESTOR_NAME', required: false) || '',
```
Lets add a filler here, we can leave the others as empty strings
```suggestion
          requestor_name: row_value(row, field: 'REQUESTOR_NAME', required: false) || 'LINK User',
```

gigxz reviewed (commented):
Ok, my second review round is done. This looks great 💯 
See comments for a few requested changes.

ttoomey committed:
Merge branch 'release-86' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers address pr feedback

* fix unit occupancy user_id value
* include full row details on missing value exception
* log missing unit assignments, surface file/line number to csv row
* remove reference to ProjectsImportAttempt
* esg funding assistance fixes
* fix enrollment coc on referral posting

ttoomey committed:
185495405-custom-data-importers placeholder value

ttoomey committed:
185495405-custom-data-importers test behavior fix

ttoomey committed:
185495405-custom-data-importers rubocop

ttoomey commented:
Thanks @gigxz! I know this PR got kind of large, I appreciate the detailed review

I think I've addressed all the feedback

Some things I noticed on the client address:
* Address county. Looks like we don't have a county field on customClientAddress. We have a district field but that might have some other significance. 
* Address country is not provided. Perhaps we should default country to 'USA'?  Or maybe that's assumed if it's empty
* I noticed we were't handling all state names in the loader. I added in a state name => code mapping to handle that

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb:
```diff
@@ -0,0 +1,151 @@
+      value
+    end
+
+    def system_user_id
+      system_user.user_id
+    end
+
+    def system_user_pk
+      system_user.id
+    end
+
+    def system_user
+      @system_user ||= Hmis::Hud::User.system_user(data_source_id: data_source.id)
+    end
```
The places where we store a primary key for user id (non-HUD tables) need to reference the `users` table. Maybe the below is clearer. We never reference the pk on the HUD user table (afaik).

```suggestion
    def system_hud_user_id
      system_hud_user.user_id
    end

    def system_user_pk
      system_user.id
    end

    def system_hud_user
      @system_hud_user ||= Hmis::Hud::User.system_user(data_source_id: data_source.id)
    end

    def system_user
      @system_user ||= Hmis::User.system_user
    end
```

gigxz reviewed (changes requested):
sorry for the delayed re review! looks great, just one change needed to store accurate user id on the unit occupancy table

gigxz commented:
Oh and county should go into district. it's just called "district" because shape is copied from FHIR https://build.fhir.org/datatypes.html#Address

gigxz committed:
Fix system_user_pk

gigxz committed:
Store county

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,274 @@
+          mci_id = row_value(member_row, field: 'MCI_ID')
+          enrollment = Hmis::Hud::Enrollment.new(default_attrs)
+          enrollment.attributes = {
+            personal_id: mci_id,
```
@ttoomey I missed this the first time, it's not quite right. The Personal ID of the enrollment needs to match `Client.PersonalID` of the client record. So really its `Hmis::Hud::Client.find(client_pk_by_mci_id(mci_id)).personal_id`. Not sure if there is already a mapping like that (mci => personal id). Can you update?

ttoomey replied:
Capturing discussion with @gigxz on this: The client Personal ID is the "MCI unique ID", not the "MCI ID". The column in the sheet is the "MCI ID" so the mapping isn't direct, need to go through external IDs instead.
* build mapping from MCI ID using external ids
* fix other importers that map MCI ID => Personal ID

ttoomey replied:
Corrected this, looks like the referrals were the only impart impacted (the client address / contact points use personal ids)

ttoomey committed:
185495405-custom-data-importers

* fix MCI to personal id mapping for referrals
* tolerate missing enrollment and client ids in importers

ttoomey committed:
Merge branch 'release-86' into 185495405-custom-data-importers

ttoomey committed:
185495405-custom-data-importers backout changes

ttoomey committed:
185495405-custom-data-importers test fix

ttoomey committed:
185495405-custom-data-importers bugfixes

ttoomey commented:
* Adjusted importers to tolerate missing MCI / enrollment IDs
* correct mapping from MCI ID to personal Id
* backed out unnecessary changes

gigxz reviewed (approved):
looks great! 🌟

ttoomey committed:
185495405-custom-data-importers minor fixes

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb:
```diff
@@ -0,0 +1,115 @@
+          record.custom_data_elements.build(attr)
+        end
+        record
+      end
```
```suggestion
      end.compact
```

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb:
```diff
@@ -0,0 +1,161 @@
+
+    def cde_definition(owner_type:, key:)
+      @cache ||= {}
+      @cache[[owner_type, key]] ||= cde_definitions.find_or_create(owner_type: owner_type, key: key)
```
I'm getting an error here that I can't figure out. I thought it might be that the User Id was different, but its the same.
```
[/tmp/s3-2023-08-22/custom-data/FederalPovertyLevel.csv:0] ActiveRecord::RecordNotUnique PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "unique_index_ensuring_one_key_per_record_type"
DETAIL:  Key (owner_type, key)=(Hmis::Hud::IncomeBenefit, federal_poverty_level) already exists.
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_definitions.rb:22:in `find_or_create'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/base_loader.rb:47:in `cde_definition'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_base_loader.rb:35:in `new_cde_record'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:39:in `block in build_records'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:32:in `block in each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:31:in `with_index'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/csv_file.rb:31:in `each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:25:in `map'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/federal_poverty_level_loader.rb:25:in `build_records'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/custom_data_element_base_loader.rb:12:in `perform'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:69:in `run_loader'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:44:in `block (2 levels) in run!'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:38:in `each'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:38:in `block in run!'
/app/drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/custom_data_elements_importer.rb:37:in `run!
```


and the existing record was below. I got it work by deleted the old record _and_ changing its key (since uniqueness constraint considers deleted records too I guess), so probably not a big deal, but just wanted to flag it.
```
Hmis::Hud::CustomDataElementDefinition.where(key: :federal_poverty_level)
=>
[#<Hmis::Hud::CustomDataElementDefinition:0x00007f4ba37af648
  id: 14,
  owner_type: "Hmis::Hud::IncomeBenefit",
  custom_service_type_id: nil,
  field_type: "string",
  key: "federal_poverty_level",
  label: "Federal Povery Level",
  repeats: false,
  data_source_id: 21,
  UserID: "1",
  DateCreated: Thu, 10 Aug 2023 13:56:13.604071000 EDT -04:00,
  DateUpdated: Thu, 10 Aug 2023 13:56:13.604071000 EDT -04:00,
  DateDeleted: nil,
  at_occurrence: false>]
```

ttoomey replied:
Oh yeah, that makes sense. Seems like we should adjust the unique constraint to exclude deleted records. This could come up at runtime too if someone removes and re-creates a CDE

ttoomey replied:
tracked this here:
https://www.pivotaltracker.com/story/show/185894590

gigxz replied:
Definitely. The issue was coming up even _before_ I deleted the record, though, and I didn't understand why.

gigxz replied:
This was because of a typo in the label on the server (Federal "Povery" Level)

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/reason_for_exit_loader.rb:
```diff
@@ -0,0 +1,90 @@
+        exit_id = row_value(row, field: 'ExitID')
+        owner_id, enrollment_id = exit_lookup[exit_id]
+
+        unless exit_id
+          log_skipped_row(row, field: 'ExitID')
+          next # early return
+        end
+
+        raise 'ExitID/EnrollmentID mismatch' if enrollment_id != row_value(row, field: 'EnrollmentID')
```
```suggestion
        unless owner_id
          log_skipped_row(row, field: 'ExitID')
          next # early return
        end

        raise 'ExitID/EnrollmentID mismatch' if enrollment_id != row_value(row, field: 'EnrollmentID')
```

Right? I'm getting an `ExitID/EnrollmentID mismatch` error, but I think its really just that the exit id wasnt found.

ttoomey replied:
Yes, that's right. Sorry about that!

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,311 @@
+    #
+    def household_member_rows_by_referral(referral_id)
+      @household_member_rows_by_referral ||= household_member_rows.group_by { |row| row_value(row, field: 'REFERRAL_ID') }
+      @household_member_rows_by_referral.fetch(referral_id)
```
@ttoomey I'm getting a KeyError here and having a hard time following what's going on. What does it mean here if the referral id isn't found?

ttoomey replied:
It seems like this must be a referral id that has no household members in the data set itself. So bad data I guess? I'll adjust it to return an empty set in that case.

ttoomey committed:
185495405-custom-data-importers

* handle referral rows with no members
* bugfixes

gigxz committed:
Update drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb

gigxz reviewed (changes requested):
For referral postings file, handle case where we get one row per household member. ("duplicate" rows that differ only by enrollment id)

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/esg_funding_assistance_loader.rb:
```diff
@@ -0,0 +1,115 @@
+    def perform
+      records = build_records
+      # destroy existing records and re-import
+      model_class.where(data_source: data_source).destroy_all if clobber
```
@eanders @ttoomey how can we bypass acts_as_paranoid here, so that records actually get wiped rather than marked as deleted?

I just realized this clobber is just marking records as deleted. Given we're planning to do this migration in prod 3X, we don't want those older delete records to be present in the db at all.

gigxz replied:
If it needs to bypass the `destroy` hooks, I _think_ thats OK. We need to clobber most of the Custom Data Elements before running this anyway.

eanders replied:
You can call `really_destroy!`
https://github.com/rubysherpas/paranoia#usage

gigxz replied:
OK, I think we should do that.

ttoomey replied:
Probably delete_all is the way to go for a record set as long as there's no dependents

ttoomey committed:
185495405-custom-data-importers skip duplicate referral ids

ttoomey committed:
185495405-custom-data-importers

* use enrollment id to assign occupancy for accepted referrals

ttoomey committed:
185495405-custom-data-importers

hard-delete clobbered records

ttoomey committed:
185495405-custom-data-importers

close unit occupancy for exited enrollments

ttoomey commented:
@gigxz 
I pushed up a few fixes:
* skip referrals/ postings for duplicate referral ids
* hard-delete clobbered records

Also fixed what I think were some logical errors. Do these changes seem right:
* On referrals, I wasn't using the new enrollment ID col to locate the enrollment. This change is only for accepted enrollments were we create occupancy records. The importer now uses the enrollment id to create occupancies (rather than MCI on the members table) b4883eb02086c195380b08f4d280b001808702db
* When creating occupancy records, the occupancy now gets an end-date equal to the exit date of the referral (or null if the enrollment is active) commit 97270ae17678b92dbf5e893f68d52efef05a2873

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/deferred_project_unit_occupancy_loader.rb:
```diff
@@ -0,0 +1,55 @@
+        record = model_class.new(unit_id: unit_id, enrollment_id: enrollment_id)
+        record.build_occupancy_period(
+          start_date: start_date || today,
+          end_date: exit_dates_by_pk[enrollment_id],
```
Ok, this makes sense. I wasn't thinking that any of these enrollments would actually be exited, but we should account for it if it happens, and what you have here looks correct. Thank you!

Correct me if I'm wrong - there are (3) scenarios where units are assigned:

1. New Enrollments for **accepted pending** referrals –– these enrollments are all open (Exit date = null)
2. Existing Enrollments for **accepted referrals** –– these enrollments should all be open at time of csv generation, because "accepted" enrollments are necessarily open. Once all clients exit, the referral status changed to "Closed"
    * Its possible that they are indeed sending some exited enrollments in the referral file, if some members are exited but others are open. I'm not sure. But what you have here is correct.
3. Enrollments in WalkInUnitTypes –– these enrollments should all be open at time of csv generation (Exit date = null)

ttoomey replied:
Yes, that's all accurate except:
> New Enrollments for accepted pending referrals –– these enrollments are all open

No, we don't create unit occupancies for accepted pending referrals. We do create enrollments though using the MCI / Project ID. Should we create occupancies for these too? Sorry if I misunderstood

gigxz replied:
As far as I know, we should be assigning Accepted Pending enrollments to units. That's the agreed upon behavior in the application, so it should hold for imported data too. That's totally my bad I didn't specify it. PR: https://github.com/greenriver/hmis-warehouse/pull/3417

gigxz commented on drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/loaders/referral_postings_loader.rb:
```diff
@@ -0,0 +1,323 @@
+        end
+
+        unit_type_mper_id = row_value(posting_row, field: 'UNIT_TYPE_ID')
+        unit_id = assign_next_unit(
+          enrollment_pk: enrollment_pk,
+          unit_type_mper_id: unit_type_mper_id,
+          fallback_start_date: parse_date(row_value(posting_row, field: 'STATUS_UPDATED_AT')),
+        )
```
This will assign every member of the household to separate units.
In reality, they should all be assigned to the same unit. I didn't think about that, sorry.  I'm pushing the change!

gigxz replied:
https://github.com/greenriver/hmis-warehouse/pull/3358/commits/b5faac9412bf9075ad2b7098297f715d2a5740b4

ttoomey replied:
Oh shoot, sorry I missed that! Thanks for fixing!

gigxz replied:
I ended up backing out this change actually https://github.com/greenriver/hmis-warehouse/pull/3358/commits/1175a328e6fd46ceb224b751aa989d72536bf9e4  because I realized that you did already do it! The `unit_pk_for_enrollment_pk` function in the `ProjectUnitTracker` appears to return the unit that's already been assigned to that household, if there is one. Can you check me on that? sorry, chaotic over here 🌀

gigxz committed:
Assign household members to the same unit

gigxz commented:
Yep @ttoomey the new changes you made look good to me. Thank you!  I'm going to try running this again against the whole data set this morning.

gigxz committed:
Bug fixes in referral

gigxz committed:
Loader updates

gigxz committed:
Fix and update test

gigxz committed:
Fix bug in MCI Mapping Importer, and other job/logging updates (#3409)

gigxz committed:
Merge branch 'release-86' into 185495405-custom-data-importers

gigxz commented:
@ttoomey I pushed some fixes
https://github.com/greenriver/hmis-warehouse/pull/3358/commits/3416a07151bd9f3588816d59763f772b97dcf5b1 - the fixes we made when pairing
https://github.com/greenriver/hmis-warehouse/pull/3358/commits/326263db0c5bda86d702c11fa1bf6d767bb26c8d - issues that came up after. most of these are making the importer more able to handle bad data. The scenarios encountered are explained in the email I just sent

gigxz commented:
Will merge once tests pass

gigxz committed:
Merge branch 'release-86' into 185495405-custom-data-importers

gigxz committed:
Comment out dependent destroy and link bug

ttoomey committed:
185495405-custom-data-importers

* delete all wip enrollments if clobber
* possible fix for missing stack trace on exceptions
* don't crash if enrollment.wip
* fix n+1 in project tracker

gigxz committed:
Merge pull request #3416 from greenriver/tt/185495405-custom-data-importers

185495405-custom-data-importers

gigxz committed:
Revert household change that was unnecessary

gigxz committed:
Merge branch 'release-86' into 185495405-custom-data-importers

gigxz committed:
Assign Accepted Pending (WIP) enrollments to units

gigxz committed:
Handle exited enrollments in WalkIn file

gigxz committed:
Temp project exclusion

ttoomey committed:
185495405 assign occupancy for accepted-pending referrals

gigxz commented:
Tests are failing because of a bad warehouse structure file, it should be fixed by merging release-86 in.

I say we merge https://github.com/greenriver/hmis-warehouse/pull/3417 into this, then squash merge the whole thing into release-86. Are we ready for that @ttoomey ?

ttoomey committed:
185495405 specs check assigned unit

ttoomey commented:
> I say we merge https://github.com/greenriver/hmis-warehouse/pull/3417 into this, then squash merge the whole thing into release-86. Are we ready for that @ttoomey ?

@gigxz that sounds good! Feel free to merge, otherwise I can take care of it later tonight.

ttoomey committed:
Merge branch 'release-86' into 185495405-custom-data-importers

ttoomey committed:
Merge pull request #3417 from greenriver/gig/185495405-custom-data-importers-accepted-pending-assignment

Assign Accepted Pending (WIP) enrollments to units

ttoomey committed:
185495405-custom-data-importers referral spec

ttoomey committed:
185495405-custom-data-importers test fix
